tests/internal/test_wrapping.py

```diff
@@ -491,3 +493,198 @@ def f(a, b, c=None):
 
     assert closure(1, 2, 3) == (1, 2, 3, 42)
     assert channel == [((42,), {}), closure, ((1, 2, 3), {}), (1, 2, 3, 42)]
+
+
+NOTSET = object()
+
+
+class DummyWrappingContext(WrappingContext):
+    def __init__(self, f):
+        super().__init__(f)
+
+        self.entered = False
+        self.exited = False
+        self.return_value = NOTSET
+        self.exc_info = None
+        self.frame = None
+
+    def __enter__(self, frame):
+        self.entered = True
+        self.frame = frame
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.exited = True
+        if exc_value is not None:
+            self.exc_info = (exc_type, exc_value, traceback)
+
+    def __return__(self, value):
+        self.return_value = value
+        return value
+
+
+def test_wrapping_context_happy():
+    def foo():
+        return 42
+
+    wc = DummyWrappingContext(foo)
+    wc.wrap()
+
+    assert foo() == 42
+
+    assert wc.entered
+    assert wc.return_value == 42
+    assert not wc.exited
+    assert wc.exc_info is None
+
+    assert wc.frame.f_code.co_name == "foo"
+    assert wc.frame.f_code.co_filename == __file__
+
+
+def test_wrapping_context_unwrapping():
+    def foo():
+        return 42
+
+    wc = DummyWrappingContext(foo)
+    wc.wrap()
+    assert _UniversalWrappingContext.is_wrapped(foo)
+
+    wc.unwrap()
+    assert not _UniversalWrappingContext.is_wrapped(foo)
+
+    assert foo() == 42
+
+    assert not wc.entered
+    assert wc.return_value is NOTSET
+    assert not wc.exited
+    assert wc.exc_info is None
+
+
+def test_wrapping_context_exc():
+    def foo():
+        raise ValueError("foo")
+
+    wc = DummyWrappingContext(foo)
+    wc.wrap()
+
+    with pytest.raises(ValueError):
+        foo()
+
+    assert wc.entered
+    assert wc.return_value is NOTSET
+    assert wc.exited
+
+    _type, exc, _ = wc.exc_info
+    assert _type == ValueError
+    assert exc.args == ("foo",)
+
+
+def test_wrapping_context_exc_on_exit():
+    class BrokenExitWrappingContext(DummyWrappingContext):
+        def __exit__(self, exc_type, exc_value, traceback):
+            super().__exit__(exc_type, exc_value, traceback)
+            raise RuntimeError("broken")
+
+    def foo():
+        raise ValueError("foo")
+
+    wc = BrokenExitWrappingContext(foo)
+    wc.wrap()
+
+    with pytest.raises(RuntimeError):
```

Comment:
should we suppress these? is there an instance where we want our exception to raise?

Reply:
This is another place where subclasses could help I guess... we can make a `RobustWrappingContext` that can never crash or fail on the wrapped call (lofty request, but maybe)